### ISSUE TITLE
New apply_linter()

### ISSUE THREAD

**Issue by Bisaloo** [r-lib/lintr#2886]
I commonly see

```r
apply(x, 2, is.numeric)
```

https://github.com/search?q=org%3Acran+%2F%5Cbapply%5C%28.*%2C+is.numeric%5C%29%2F&type=code

But I think this is often a mistake overlooking the fact that:

1. `apply()` will coerce the input to a matrix
2. matrix can only have one type

I believe this should be linted to suggest either `vapply(x, seq_len(ncol(x)), function(i) is.numeric(x[[i]])`, or simply `is.numeric(x)` depending on the desired check.

A similar situation occurs with `apply(x, 2, is.character)`, `apply(x, 2, is.integer)`, etc. and `apply(x, 2, typeof)`.

Am I missing a situation where it makes sense to use these patterns?

**Comment by MichaelChirico** [comment]
Good point... Maybe we should rename `matrix_apply_linter` --> `apply_linter` and include both rules?

**Comment by AshesITR** [comment]
Is there any reasonable use case for `is.numeric(as.matrix(x))`? `as.matrix()` is potentially very expensive and the output type of `as.matrix(x)` seems rather predictable to me based on the column types of `x`.

If I'm not missing anything, I'd suggest something along the lines of "Use ... to check column types. apply(...) is likely a mistake because...".

**Comment by MichaelChirico** [comment]
I mean, I guess in principle you could be checking that the column types in the data.frame have `max_type==numeric`? Even then it's a very inefficient way to do so. But if you know your matrix is tiny it's probably the cleanest approach (vs something like `all(vapply(x, typeof) %in% c("raw", "logical", "integer", "double"))` ?)

**Comment by Bisaloo** [comment]
An usage I have seen that might be replaced by `is.numeric(as.matrix(x))` is in:

```r
if (!all(apply(x, 2, is.numeric)) {
  stop("x needs to contain only numeric")
}
```

It doesn't work for reasons the authors probably expected but it works because if there is a character column somewhere, `x` will be coerced to a character matrix.

Using `vapply()` might still be more efficient, just slightly more verbose.

**Comment by Bisaloo** [comment]
There is also a good amount of cases where `x` is already a matrix and the correct replacement is simply `is.numeric(x)`.
